llvm/test/CodeGen/AMDGPU/xor3-i1-const.ll

```diff
@@ -9,7 +9,9 @@ define amdgpu_ps float @xor3_i1_const(float inreg %arg1, i32 inreg %arg2) {
 ; GCN-NEXT:    v_cmp_lt_f32_e64 s[2:3], s0, 0
 ; GCN-NEXT:    v_cmp_lt_f32_e32 vcc, s0, v0
 ; GCN-NEXT:    s_and_b64 s[0:1], s[2:3], vcc
-; GCN-NEXT:    v_cndmask_b32_e64 v0, 1.0, 0, s[0:1]
+; GCN-NEXT:    s_and_b64 s[0:1], s[0:1], exec
+; GCN-NEXT:    s_cselect_b32 s0, 0, 1.0
+; GCN-NEXT:    v_mov_b32_e32 v0, s0
```

Comment:
@alex-t codegen is worse here, because the condition came from a VALU compare _and_ the result of the select needs to be in a VGPR. Before SIFixSGPRCopies the MIR would look something like:
```
  %2:sreg_64 = nofpexcept V_CMP_EQ_F32_e64 0, %0:sgpr_32, 0, %1:sgpr_32, 0, implicit $mode, implicit $exec
  %3:sgpr_32 = S_MOV_B32 1084227584
  %4:sgpr_32 = S_MOV_B32 1056964608
  $scc = COPY %2:sreg_64 // <== copy from vector condition code
  %5:sgpr_32 = S_CSELECT_B32 killed %4:sgpr_32, killed %3:sgpr_32, implicit $scc
  $vgpr0 = COPY %5:sgpr_32 // <== copy result back to vgpr
```
After SIFixSGPRCopies it looks like:
```
  %2:sreg_64 = nofpexcept V_CMP_EQ_F32_e64 0, %0:sgpr_32, 0, %1:sgpr_32, 0, implicit $mode, implicit $exec
  %3:sgpr_32 = S_MOV_B32 1084227584
  %4:sgpr_32 = S_MOV_B32 1056964608
  %6:sreg_64 = S_AND_B64 %2:sreg_64, $exec, implicit-def $scc
  %5:sgpr_32 = S_CSELECT_B32 killed %4:sgpr_32, killed %3:sgpr_32, implicit $scc
  $vgpr0 = COPY %5:sgpr_32
```
Do you think SIFixSGPRCopies could detect this case, and instead convert S_CSELECT_B32 into V_CNDMASK_B32?

Reply:
SIFixSGPRCopies has nothing to do with that MIR at all. There are no VGPR 2 SGPR copies in the input MIR.  The only thing this pass changes is that it lowers the VCC to SCC copy to S_AND_B64. 
I agree that the Select SDNode with Cond operand produced by VALU and a user expecting the result in VGPR should be selected to V_CNDMASK_B32. Since it is not related to the VGPR to SGPR copies lowering I am curious what would be the right place for that optimization? It seems like it should be better done in the ISel.
The problem is that the selection walks the DAG from the root upward and selecting the Select SDNode we would have to walk its Cond operand sub-DAG to guess if it is going to be selected to VALU. For this particular case, we could deduce it from the fact that Setcc has an operand of FP type and we have no scalar FP comparison. So, I would think of the custom selection callback for that purpose.
I attached the selection DAG after DAGCombine
[combined.pdf](https://github.com/user-attachments/files/17385275/combined.pdf)
 to illustrate the idea.

Reply:
I realize there are no VGPR to SGPR copies, but there is a VCC to SCC copy which is morally kind of the same thing: we could use a cost model to decide whether to implement that copy (by inserting S_AND) or to promote the user of scc to an equivalent VALU op.

It's hard to do this at instruction selection time because the optimal sequence depend on whether the result of select needs to go in an SGPR or not:
```
  %2 = V_CMP_EQ_F32_e64
  %5 = V_CNDMASK_B32
  ; result is in VGPR
```
vs:
```
  %2 = V_CMP_EQ_F32_e64
  %6 = S_AND_B64
  %5 = S_CSELECT_B32
  ; result is in SGPR
```
In the second case we don't want to use V_CNDMASK_B32 because we would need to insert V_READFIRSTLANE_B32 afterwards which is generally bad for performance.

Reply:
Yeah. It make sense. I will look how to change lowerSCCCopies to achieve what we want.

Reply:
What I am concerned about is that there might be other users of the SCC definition introduced by COPY.
In the general case, we would have to DFS-walk looking for a definition that kills the current one or other users to decide if it is possible to remove the copy and convert s_cselect_b32 to v_cndmask.

Reply:
I sketched out a change that does what we want but it does not take care of the general case. Although, it produces a good asm for your example :)       
       v_mov_b32_e32 v0, 0x42640000
        v_cmp_lt_f32_e64 s[2:3], s0, 0
        v_cmp_lt_f32_e32 vcc, s0, v0
        s_and_b64 s[0:1], s[2:3], vcc
        v_cndmask_b32_e64 v0, 1.0, 0, s[0:1]
I will create a PR for the change so that you can look at the code.

Reply:
> What I am concerned about is that there might be other users of the SCC definition introduced by COPY.
> In the general case, we would have to DFS-walk looking for a definition that kills the current one or other users to decide if it is possible to remove the copy and convert s_cselect_b32 to v_cndmask.

Don't you have the same problem with a regular VGPR to SGPR copy? Do you analyse all uses of the SGPR in the whole function? Or it is different because SCC is a physical register?

Reply:
> > What I am concerned about is that there might be other users of the SCC definition introduced by COPY.
> > In the general case, we would have to DFS-walk looking for a definition that kills the current one or other users to decide if it is possible to remove the copy and convert s_cselect_b32 to v_cndmask.
> 
> Don't you have the same problem with a regular VGPR to SGPR copy? Do you analyse all uses of the SGPR in the whole function? Or it is different because SCC is a physical register?

Yes, SCC is a physreg so we cannot rely on SSA looking for its uses. That is my concern.

Reply:
I created https://github.com/llvm/llvm-project/pull/113705 as a discussion board.